### DISCUSSION TITLE
Use Emulator

### DISCUSSION THREAD

**Discussion by FMGordillo** [gladly-team/next-firebase-auth#63] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
When you have local/staging/prod environments, it gets quite tricky to test your changes (specially if you have Cloud Functions that handles the user creation (`functions.auth.user().onCreate()`).

**Describe the solution you'd like and how you'd implement it**
Currently `react-firebaseui` supports something like this:

```
        <StyledFirebaseAuth
          uiConfig={firebaseAuthConfig}
          firebaseAuth={
            process.env.NODE_ENV !== "production"
              ? firebase.auth().useEmulator("http://localhost:9099/")
              : firebase.auth()
          }
        />
```

But when I tried to use it, it breaks somewhere else:
![image](https://user-images.githubusercontent.com/7970174/106360782-3de92380-62f9-11eb-8e06-86ab0461958c.png)

**Is this a breaking change?**
I don't think so

**Describe alternatives you've considered**
I have no idea if there's any alternative

**Comment by kmjennison** [comment]
You can still initialize Firebase yourself (both JS SDK and admin SDK), before you call `initAuth`. When a Firebase app exists, this package will use it. Could you use the emulator that way?

**Reply by Parnswir** [reply]
I tried. Most of it is working fine, however, calling `setAuthCookies` (within the `/api/login` route) will fetch a hardcoded Google API ([here](https://github.com/gladly-team/next-firebase-auth/blob/aaa15e0207d0355fcfef77c7db786344347020a5/src/firebaseAdmin.js#L93)) which fails.

**Reply by kmjennison** [reply]
Got it. Do you have suggestions on how to emulate that server-side API behavior, or do Google/Firebase have any tools that might help? Happy to see how we can make this package compatible, if possible.

**Comment by Parnswir** [comment]
I don't. I have been using firebase and next-firebase-auth for a week now,
so I'd need to dig into it.

The emulator is under pretty active development and it seems validating jwt
was added only two weeks ago. The docs aren't reflecting this yet, however.

One step towards emulator compatibility would be to offer setting a flag in
the config and then piping that to the initializeApp calls. However, the
admin SDK is a bit picky about this and basically wants it to be set as
environment variable. I can provide a sample app in the coming days.

**Comment by Parnswir** [comment]
Quick and dirty, it could look like this: https://github.com/gladly-team/next-firebase-auth/compare/main...Parnswir:emulator-poc?expand=1

To start the emulator, download the firebase CLI, cd to the example directory of the repo and run `firebase emulators:start`.

- I have initialized the Firebase Client SDK and the Admin SDK by myself, before next-firebase-auth.
- The Client SDK configuration is ignored once you use `auth.useEmulator("http://localhost:9099");`.
- The Admin SDK configures itself automatically with `process.env.FIREBASE_AUTH_EMULATOR_HOST="localhost:9099"`

The basics work, but then the errors start rolling in. Some are probably connected to the example app, but as described above, setAuthCookie will always call the (live) identitykit instead of the emulator.

**Comment by litewarp** [comment, marked as answer]
Just an FYI - I was able to get server-side refresh tokens working with the emulator by tweaking the verification url.  I just submitted a PR to this effect. Hopes this resolves the issue!

**Reply by FMGordillo** [reply]
Just a quick update (sorry for the delay on that)

https://github.com/litewarp/next-firebase-auth/pull/1

It is merged! I have not cheched if it works, but I trust you in this one. Thank you!
